### PR TITLE
Workaround for hangs of the docker_client in the facts genereation

### PR DESCRIPTION
**Pull request by mgrzybowski**
Hi
  From time to time puppet agent is hanging indefinitely on **docker** client command  :

>  ├─puppet,2645 /usr/bin/puppet agent
  │   ├─puppet,2652
  │   │   ├─sh,2766 -c /usr/bin/docker version --format '{{json .}}'
  │   │   │   └─docker,2767 version --format {{json .}}
  │   │   │       ├─{docker},2776
  │   │   │       ├─{docker},2777
  │   │   │       ├─{docker},2778
  │   │   │       ├─{docker},2779
  │   │   │       ├─{docker},2780
  │   │   │       ├─{docker},2783
  │   │   │       ├─{docker},2784
  │   │   │       ├─{docker},2785
  │   │   │       ├─{docker},2786
  │   │   │       ├─{docker},2787
  │   │   │       ├─{docker},2788
  │   │   │       ├─{docker},2789
  │   │   │       ├─{docker},2792
  │   │   │       ├─{docker},2793
  │   │   │       ├─{docker},2816
  │   │   │       ├─{docker},2817
  │   │   │       ├─{docker},6331
  │   │   │       ├─{docker},6332
  │   │   │       ├─{docker},6333
  │   │   │       ├─{docker},7149
  │   │   │       ├─{docker},7150
  │   │   │       ├─{docker},9269
  │   │   │       ├─{docker},9270
  │   │   │       ├─{docker},32457
  │   │   │       ├─{docker},14309
  │   │   │       ├─{docker},16658
  │   │   │       ├─{docker},17687
  │   │   │       ├─{docker},20057
  │   │   │       ├─{docker},24652
  │   │   │       ├─{docker},28391
  │   │   │       ├─{docker},29383
  │   │   │       ├─{docker},12255
  │   │   │       ├─{docker},22690
  │   │   │       ├─{docker},15546
  │   │   │       ├─{docker},11915
  │   │   │       └─{docker},20370
  │   │   └─{puppet},2654

putting some kind of time limit in docker clinet command will prevent this to happen .

( i suspect systemd socket handling when docker deamon is restarted/upgraded is  main cause,
but i was not able to reporduce this )

### CODE REVIEW

**Comment by sheenaajay** [comment]
@mgrzybowski Thanks for submitting the PR. Could you try adding the timeout to facts. Example here https://puppet.com/docs/facter/3.7/custom_facts.html#execution-timeouts.

**Comment by mgrzybowski** [comment]
Hi Sheenaajay

I'm simulating broken docker socket by disabling  docker service and creating dummy socket:

`nc -lkU  /var/run/docker.sock`

Then i'm executing:

`facter  -p  docker_version`

Changing **Facter::Util::Resolution.exec** to **Facter::Util::Resolution.execute** and adding the  timeout option is not working for me . Facter is hanging indefinitly.


>  ├─sshd -D
>  │   ├─sshd    
>  │   │   └─bash
>  │   │       └─facter /usr/bin/facter -p docker_version
>  │   │           ├─sh -c /usr/bin/docker version --format '{{json .}}'
>  │   │           │   └─docker version --format {{json .}}
>  │   │           │       └─8*[{docker}]
>  │   │           └─{ruby-timer-thr}


exact code snippet that was testing:
```
        docker_json_str = Facter::Util::Resolution.execute(
          "#{docker_command} info --format '{{json .}}'",
          options = {:timeout => 30},
        )
```

Am I missing something here?

**Comment by DavidS** [comment]
@mgrzybowski The quoted section in the docs reads `Facter::Core::Execution.execute('sleep 10', options = {:timeout => 5})`, not `Facter::Util::Resolution.execute`. I tried that with the example given there and that did work:
```
david@davids:~/tmp$ cat facts/sleep.rb 
Facter.add(:sleep) do
  setcode do
    begin
      Facter::Core::Execution.execute('sleep 10', options = {:timeout => 5})
      'did not timeout!'
    rescue Facter::Core::Execution::ExecutionFailure
      'timeout!'
    end
  end
end

david@davids:~/tmp$ date; facter --custom-dir ./facts/ sleep; date
Tue 17 Dec 17:39:12 GMT 2019
timeout!
Tue 17 Dec 17:39:17 GMT 2019
david@davids:~/tmp$ 
```
can you confirm that this example is working for you and that the same fact code calling to docker instead is not working?

**Comment by mgrzybowski** [comment]
Hi
  Since Facter::Util::Resolution is extended by Core::Execution (https://www.rubydoc.info/gems/facter/Facter/Util/Resolution)  i was sure that this should work the same.

My facter have to be to old for execute **timeout** option to work:
```
# cat  /var/lib/puppet/lib/facter/sleep.rb 
require 'facter'
Facter.add(:sleep) do
  setcode do
    begin
      Facter::Core::Execution.execute('sleep 10', options = {:timeout => 5})
      'did not timeout!'
    rescue Facter::Core::Execution::ExecutionFailure
      'timeout!'
    end
  end
end
```

```
# date; facter -p sleep; date
Mon Dec 23 14:49:23 UTC 2019
did not timeout!
Mon Dec 23 14:49:33 UTC 2019
```

```
# facter  -v
2.4.6
# facter  -p os 
{"name"=>"Ubuntu", "family"=>"Debian", "release"=>{"major"=>"16.04", "full"=>"16.04"}, "lsb"=>{"distcodename"=>"xenial", "distid"=>"Ubuntu", "distdescription"=>"Ubuntu 16.04.6 LTS", "distrelease"=>"16.04", "majdistrelease"=>"16.04"}}
```

I barely know ruby but looking at the gem source: https://www.rubydoc.info/gems/facter/Facter/Core/Execution/Base
i can not see how this **timeout** option should work since the command is executeded by statement
```
out = %x{#{expanded_command}}.chomp
```

**Comment by sheenaajay** [comment]
@mgrzybowski Apologies for the delay. timeout is facter 3 feature. we retested them and its all working for both scenarios.
`[root@inert-alligator lib]# facter sleep --custom-dir=./facter 
timeout!
`
`[root@inert-alligator lib]# facter sleep --custom-dir=./facter 
did not timeout!
`

We will get the changes in. Thank you

File: lib/facter/docker.rb [hunk 1]
```diff
@@ -69,7 +69,7 @@ def interfaces
   setcode do
     if Facter::Util::Resolution.which('docker')
       value = Facter::Util::Resolution.exec(
-        "#{docker_command} version --format '{{json .}}'",
+        "#{docker_command} version --format '{{json .}}'",  timeout: 90
```
Comment (sheenaajay) [inline lib/facter/docker.rb:72]:
```suggestion
        "#{docker_command} version --format '{{json .}}'", timeout: 90
```

File: lib/facter/docker.rb [hunk 2]
```diff
@@ -69,7 +69,7 @@ def interfaces
   setcode do
     if Facter::Util::Resolution.which('docker')
       value = Facter::Util::Resolution.exec(
```
Comment (sheenaajay) [inline lib/facter/docker.rb:71]:
```suggestion
      value = Facter::Util::Execution.exec(
```

File: lib/facter/docker.rb [hunk 3]
```diff
@@ -83,20 +83,20 @@ def interfaces
     if docker_version !~ %r{1[.][0-9][0-2]?[.]\w+}
       if Facter::Util::Resolution.which('docker')
         docker_json_str = Facter::Util::Resolution.exec(
```
Comment (sheenaajay) [inline lib/facter/docker.rb:85]:
```suggestion
        docker_json_str = Facter::Util::Execution.exec(
```

File: lib/facter/docker.rb [hunk 4]
```diff
@@ -83,20 +83,20 @@ def interfaces
...
-          "#{docker_command} info --format '{{json .}}'",
+          "#{docker_command} info --format '{{json .}}'", timeout: 90
         )
         begin
           docker = JSON.parse(docker_json_str)
           docker['network'] = {}
 
           docker['network']['managed_interfaces'] = {}
-          network_list = Facter::Util::Resolution.exec("#{docker_command} network ls | tail -n +2")
+          network_list = Facter::Util::Resolution.exec("#{docker_command} network ls | tail -n +2", timeout: 90)
```
Comment (sheenaajay) [inline lib/facter/docker.rb:93]:
```suggestion
          network_list = Facter::Util::Execution.exec("#{docker_command} network ls | tail -n +2", timeout: 90)
```

File: lib/facter/docker.rb [hunk 5]
```diff
@@ -83,20 +83,20 @@ def interfaces
...
           docker_network_names = []
           network_list.each_line { |line| docker_network_names.push line.split[1] }
           docker_network_ids = []
           network_list.each_line { |line| docker_network_ids.push line.split[0] }
           docker_network_names.each do |network|
-            inspect = JSON.parse(Facter::Util::Resolution.exec("#{docker_command} network inspect #{network}"))
+            inspect = JSON.parse(Facter::Util::Resolution.exec("#{docker_command} network inspect #{network}"), timeout: 90)
```
Comment (sheenaajay) [inline lib/facter/docker.rb:99]:
```suggestion
            inspect = JSON.parse(Facter::Util::Execution.exec("#{docker_command} network inspect #{network}"), timeout: 90)
```

**Comment by sheenaajay** [comment]
@mgrzybowski Sorry for the delay. Don't have permissions to push changes to your branch. Could you please provide permissions or I can create a new PR from this. Thank you again.

**Comment by mgrzybowski** [comment]
Hi @sheenaajay . You should receive invite from github . Plese check and let me know if it did not work .

**Comment by sheenaajay** [comment]
@mgrzybowski Perfect. Thank you.

**Comment by david22swan** [comment]
@mgrzybowski 
Making a quick change to point this PR towards the newly defined default main branch as part of our work to remove inappropriate terminology. Apologies if this is inconvenient in any way.

**Comment by sanfrancrisko** [comment]
Given the dependency on `facter >= 3`, this will not work on Puppet `5` - we'll need to hold off on this merge until we can bump the dependency of the module to `Puppet >= 6`

**Comment by adrianiurca** [comment]
Hey @mgrzybowski, can you rebase the PR in order to test using the latest workflow?

kind regards,
@adrianiurca

**Comment by saz** [comment]
Any news here? Running into the issue and it's a bit annoying :-)

**Pull request closed without merging**